File: week_1/day_1_arrays/kadane_follow_up_questions/Modify_this_for_circular_arrays.js
function maxSubarrayCircular(nums) {
    let totalSum = 0;
    let maxSum = nums[0], currMax = nums[0];
    let minSum = nums[0], currMin = nums[0];

    for (let num of nums) {
        totalSum += num;

        currMax = Math.max(num, currMax + num);
        maxSum = Math.max(maxSum, currMax);

        currMin = Math.min(num, currMin + num);
        minSum = Math.min(minSum, currMin);
    }

    if (maxSum < 0) return maxSum;

    return Math.max(maxSum, totalSum - minSum);
}

console.log(maxSubarrayCircular([1, -2, 3, -2]));

File: week_1/day_1_arrays/kadane_follow_up_questions/Modify_this_for_circular_arrays.py
def maxSubarrayCircular(nums):
    total_sum = 0
    max_sum = curr_max = nums[0]
    min_sum = curr_min = nums[0]

    for num in nums:
        total_sum += num
        curr_max = max(num, curr_max + num)
        max_sum = max(max_sum, curr_max)

        curr_min = min(num, curr_min + num)
        min_sum = min(min_sum, curr_min)

    # All numbers negative case
    if max_sum < 0:
        return max_sum

    return max(max_sum, total_sum - min_sum)

nums = [1, -2, 3, -2]
print(maxSubarrayCircular(nums))
